src/ledger/LedgerTxn.cpp

```diff
@@ -170,6 +171,82 @@ LedgerEntryPtr::isDeleted() const
 {
     return mState == EntryPtrState::DELETED;
 }
+bool
+LedgerKeyMeter::canLoad(LedgerKey const& key, size_t entrySizeBytes) const
+{
+    return maxReadQuotaForKey(key) >= entrySizeBytes;
+}
+
+void
+LedgerKeyMeter::addTxn(SorobanResources const& resources)
+{
+    mTxReadBytes.push_back(resources.readBytes);
+    auto txId = mTxReadBytes.size() - 1;
+    auto addKeyToTxnMap = [&](auto const& key) {
+        mLedgerKeyToTxs[key].emplace_back(txId);
+    };
+    std::for_each(resources.footprint.readOnly.begin(),
+                  resources.footprint.readOnly.end(), addKeyToTxnMap);
+    std::for_each(resources.footprint.readWrite.begin(),
+                  resources.footprint.readWrite.end(), addKeyToTxnMap);
+}
+
+void
+LedgerKeyMeter::updateReadQuotasForKey(LedgerKey const& key,
+                                       size_t entrySizeBytes)
+{
+    auto iter = mLedgerKeyToTxs.find(key);
+    if (iter == mLedgerKeyToTxs.end())
+    {
+        // Key does not belong to the footprint of any transaction.
+        return;
+    }
+    // Update the read quota for every transaction containing this key.
+    bool exceedsQuotaForAllTxns = true;
+    for (auto txn : iter->second)
+    {
+        auto& quota = mTxReadBytes.at(txn);
+        if (quota < entrySizeBytes)
+        {
+            quota = 0;
+        }
+        else
+        {
+            exceedsQuotaForAllTxns = false;
+            quota -= entrySizeBytes;
+        }
+    }
+    if (exceedsQuotaForAllTxns)
+    {
+        mNotLoadedKeys.insert(key);
+    }
+}
+
+uint32_t
+LedgerKeyMeter::maxReadQuotaForKey(LedgerKey const& key) const
+{
+    auto iter = mLedgerKeyToTxs.find(key);
+    if (iter == mLedgerKeyToTxs.end())
+    {
+        // Key does not belong to the footprint of any transaction,
+        // therefore it is not quota-limited.
+        return std::numeric_limits<uint32_t>::max();
```

Comment:
While there are some edge cases like source accounts that aren't in a TX footprint, we should always meter CONTRACT_CODE and CONTRACT_DATA entries. Since we removed the stronger assert due to the edge cases, it might be good to add an assert here that we're not accidentally unmetering soroban entries.